### PR TITLE
Updated Claimer Agent to support new multi-claim interface

### PR DESCRIPTION
**Pull request by asselstine**

### CODE REVIEW

File: src/ClaimerAgent.sol [hunk 1]
```diff
@@ -4,76 +4,219 @@ pragma solidity 0.8.17;
...
+        if (targetClaimCount > 0) {
+            console2.log("GO FOR CLAIMS", targetClaimCount);
+        }
+
+        if (targetClaimCount > 0) {
+            // console2.log("claimPrizes(computedDrawId, targetClaimCount)", computedDrawId, targetClaimCount);
+            uint earned = claimPrizes(targetClaimCount);
+            console2.log("CLAIMED ", targetClaimCount, " PRIZES. EARNED ", earned / 1e18);
+            console2.log("\tmin claim fee: ", claimCostInPrizeTokens/1e18);
+        }
```
Comment (PierrickGT) [inline src/ClaimerAgent.sol]: These two if can be combined.

File: src/ClaimerAgent.sol [hunk 2]
```diff
@@ -4,76 +4,219 @@ pragma solidity 0.8.17;
...
+        uint _winnersLength
+    ) public view returns (uint32[] memory prizeIndexLength) {
+        uint256 winnersLength = countWinners(_nextPrizeIndex, _claimCount);
+        prizeIndexLength = new uint32[](_winnersLength);
+        uint256 winnerIndex;
+        address currentWinner;
+        for (uint p = _nextPrizeIndex; p < _nextPrizeIndex + _claimCount; p++) {
+            if (currentWinner == address(0)) {
+                currentWinner = drawPrizes[computedDrawId][p].winner;
+            }
```
Comment (PierrickGT) [inline src/ClaimerAgent.sol:110]: Not sure this condition is needed.
For the first iteration, `currentWinner` will be equal to `address(0)` so it should also be different from `drawPrizes[computedDrawId][p].winner`.

File: src/ClaimerAgent.sol [hunk 3]
```diff
@@ -4,76 +4,219 @@ pragma solidity 0.8.17;
...
+
+        // populate result arrays
+        uint winnerIndex = 0;
+        address currentWinner;
+        uint256 prizeIndex;
+        for (uint pi = _nextPrizeIndex; pi < _nextPrizeIndex + _claimCount; pi++) {
+            if (currentWinner == address(0)) {
+                currentWinner = drawPrizes[computedDrawId][pi].winner;
+                winners[winnerIndex] = currentWinner;
             }
```
Comment (PierrickGT) [inline src/ClaimerAgent.sol]: Also not sure if this condition is needed.

File: src/ClaimerAgent.sol [hunk 4]
```diff
@@ -4,76 +4,219 @@ pragma solidity 0.8.17;
...
-            env.claimer().claimPrizes(drawId, claims, address(this));
-            console2.log("CLAIMING ", claimCount, " PRIZES");
-            draws[drawId].currentClaimIndex += claimCount;
+
+            // if the current prize is for a different winner, then skip ahead
```
Comment (PierrickGT) [inline src/ClaimerAgent.sol]: This comment is a bit misleading.
If the `currentWinner` is different from the current prize winner, we don't skip ahead but enter in the condition and update `currentWinner` with this new winner.
So maybe it should be rephrased `then skip to the next winner`.

**Review by PierrickGT — APPROVED** [review]

File: test/ClaimerAgent.t.sol [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    
+    ClaimerAgent agent;
+
+    uint numTiers = 2;
+
+    function setUp() public {
+        vm.etch(address(env), "environment");
+        vm.etch(address(prizePool), "prizePool");
+        vm.etch(address(vault), "vault");
+        vm.etch(address(claimer), "claimer");
```
Comment (trmid) [inline test/ClaimerAgent.t.sol:29]: Getting an error when running tests saying:
`[FAIL. Reason: Setup failed: Etch cannot be used on precompile addresses (N < 10). Please use an address bigger than 10 instead]`

Comment (trmid) [reply test/ClaimerAgent.t.sol:29]: forge version: `forge 0.2.0 (6defdce 2023-05-17T00:04:12.304272526Z)`

File: src/ClaimerAgent.sol [hunk 6]
```diff
@@ -4,76 +4,219 @@ pragma solidity 0.8.17;
...
+        uint256 remainingPrizes = drawPrizes[computedDrawId].length - nextPrizeIndex;
+        uint256 claimCount = targetClaimCount > remainingPrizes ? remainingPrizes : targetClaimCount;
+
+        uint totalFees;
+
+        if (claimCount == 0) {
+            return 0;
+        }
+
+        while (nextPrizeIndex < nextPrizeIndex + claimCount) {
```
Comment (trmid) [inline src/ClaimerAgent.sol]: This is a bit confusing and could be simplified.
```suggestion
        while (claimCount > 0) {
```

**Review by trmid — APPROVED** [review]

File: src/ClaimerAgent.sol (hunk 2 above)
Comment (asselstine) [reply src/ClaimerAgent.sol:110]: Trick is that we don't want winnerIndex to increment

File: src/ClaimerAgent.sol (hunk 3 above)
Comment (asselstine) [reply src/ClaimerAgent.sol]: winnerIndex!

**Pull request merged**